0002-RCBS-Crypto-building-block.md

```diff
@@ -0,0 +1,644 @@
+```
+
+### Binary header
+
+The binary header is 8-bytes long:
+
+| Cipher suite | Nonce prefix |
+| --- | --- |
+| 1 byte | 7 bytes |
+| `0x00`: AES-GCM <br>`0x01`: ChaCha20-Poly1305 ([RFC 8439](https://datatracker.ietf.org/doc/html/rfc8439)) | Random sequence of 7 bytes generated by a CSPRNG |
+
+> Dapr will choose AES-GCM if running on a system that supports AES-NI CPU instructions, or ChaCha20-Poly1305 otherwise. We may change that behavior in the future.
+
+### Segments
+
+The plaintext is chunked into segments 64KB (65,536 bytes) each; the last segment may be shorter. Segments must never be empty, unless the entire file is empty.
+
+Each segment of plaintext is encrypted independently and stored together with its tag:
+
+```
+encrypted chunk || tag
+```
+
+Segments are encrypted with a **Payload Key (PK)** that is derived from the (plain-text) File Key and the nonce prefix:
+
+```
+payload-key = HKDF-SHA-256(ikm = file key, salt = nonce prefix, info = "payload")
+```
+
+Each segment is encrypted using a different 12-byte nonce / initialization vector:
+
+```
+nonce_prefix || i || last_segment
+```
+
+Where: 
+
+- `nonce_prefix` (7 bytes) is the nonce prefix from the binary header
+- `i` (4 bytes) is the sequence number, as a 32-bit unsigned integer counter, encoded as big-endian. The first segment has sequence number 0, and it increases.
+- `last_segment` (1 byte) is `0x01` if this is the last segment, or `0x00` otherwise
+
+## Components
+
+Components in dapr/components-contrib implement low-level primitives only, while all higher-level operations are performed by the runtime, so they are executed in a consistent way across all backends/services. This is because the job of the components is limited to actually interacting with the key vaults, and everything else is best handled by the runtime.
+
+Components are to be placed in the `crypto` folder and must implement the `SubtleCrypto` interface:
+
+```go
+import "github.com/lestrrat-go/jwx/v2/jwk"
+
+// SubtleCrypto offers an interface to perform low-level ("subtle") cryptographic operations with keys stored in the vault
+type SubtleCrypto interface {
+    // GetKey returns the public part of a key stored in the vault
+    // This method returns an error if the key is symmetric
+    GetKey(
+        // Context that can be used to cancel the running operation
+        ctx context.Context,
+        // Name (or name/version) of the key to use in the key vault
+        key string,
+    ) (
+        // Object containing the public key
+        pubKey jwk.Key,
+        // Error
+        err error,
+    )
+
+    // Encrypt a small message and returns the ciphertext
+    Encrypt(
+        // Context that can be used to cancel the running operation
+        ctx context.Context, 
+        // Input plaintext
+        plaintext []byte,
+        // Encryption algorithm to use
+        algorithm string,
+        // Name (or name/version) of the key to use in the key vault
+        key string,
```

Comment:
should this be `keyName` instead ?

Reply:
This is just a name used in an internal API